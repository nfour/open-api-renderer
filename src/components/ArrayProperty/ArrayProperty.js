import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';

export default class ArrayProperty extends PureComponent {
  render() {
    const { constraints } = this.props;

    if (!constraints) {
      return null;
    }

    const { minItems, maxItems, uniqueItems } = constraints;
    const validations = [];

    if (uniqueItems) {
      validations.push('items must be unique');
    }

    if (maxItems !== undefined && minItems !== undefined) {
      // Be succint if the minItems is the same maxItems
      // ie. value can only be of `x` length.
      if (maxItems === minItems) {
        validations.push(`${minItems} items`);
      } else {
        validations.push(`${minItems}-${maxItems} items`);
      }
    } else if (minItems !== undefined) {
      validations.push(`at least ${minItems} items`);
    } else if (maxItems !== undefined) {
      validations.push(`at most ${maxItems} items`);
    }

    if (!validations.length) {
      return null;
    }

    return (
      <span>
        {validations.map(constraint =>
          <span key={constraint} className="array-constraints">{constraint}</span>
        )}
      </span>
    );
  }
}

ArrayProperty.propTypes = {
  constraints: PropTypes.shape({
    maxItems: PropTypes.number,
    minItems: PropTypes.number,
    uniqueItems: PropTypes.bool
  })
};
