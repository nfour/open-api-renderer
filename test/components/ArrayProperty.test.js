import React from 'react';
import ArrayProperty from './../../src/components/ArrayProperty/ArrayProperty';
import renderer from 'react-test-renderer';

describe('<ArrayProperty />', () => {
  it('renders nothing if there are no applicable constraints', () => {
    let tree = renderer.create(
      <ArrayProperty />
    ).toJSON();

    expect(tree).toBeNull();

    tree = renderer.create(
      <ArrayProperty constraints={{pattern: '[a-z]+'}}/>
    ).toJSON();

    expect(tree).toBeNull();
  });

  it('renders uniqueItems', () => {
    const tree = renderer.create(
      <ArrayProperty constraints={{uniqueItems: true}} />
    );

    expect(tree).toMatchSnapshot();
  });

  it('renders minItems', () => {
    const tree = renderer.create(
      <ArrayProperty constraints={{minItems: 2}} />
    );

    expect(tree).toMatchSnapshot();
  });

  it('renders maxItems', () => {
    const tree = renderer.create(
      <ArrayProperty constraints={{maxItems: 10}} />
    );

    expect(tree).toMatchSnapshot();
  });

  it('renders minItems and maxItems', () => {
    const tree = renderer.create(
      <ArrayProperty constraints={{minItems: 4, maxItems: 10}} />
    );

    expect(tree).toMatchSnapshot();
  });

  it('renders minItems and maxItems when they are the same value', () => {
    const tree = renderer.create(
      <ArrayProperty constraints={{minItems: 4, maxItems: 4}} />
    );

    expect(tree).toMatchSnapshot();
  });

  it('renders uniqueItems, minItems and maxItems', () => {
    const tree = renderer.create(
      <ArrayProperty constraints={{uniqueItems: true, minItems: 4, maxItems: 10}} />
    );

    expect(tree).toMatchSnapshot();
  });
});
