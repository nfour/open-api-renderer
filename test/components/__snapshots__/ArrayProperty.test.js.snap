// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ArrayProperty /> renders maxItems 1`] = `
<span>
  <span
    className="array-constraints"
  >
    at most 10 items
  </span>
</span>
`;

exports[`<ArrayProperty /> renders minItems 1`] = `
<span>
  <span
    className="array-constraints"
  >
    at least 2 items
  </span>
</span>
`;

exports[`<ArrayProperty /> renders minItems and maxItems 1`] = `
<span>
  <span
    className="array-constraints"
  >
    4-10 items
  </span>
</span>
`;

exports[`<ArrayProperty /> renders minItems and maxItems when they are the same value 1`] = `
<span>
  <span
    className="array-constraints"
  >
    4 items
  </span>
</span>
`;

exports[`<ArrayProperty /> renders uniqueItems 1`] = `
<span>
  <span
    className="array-constraints"
  >
    items must be unique
  </span>
</span>
`;

exports[`<ArrayProperty /> renders uniqueItems, minItems and maxItems 1`] = `
<span>
  <span
    className="array-constraints"
  >
    items must be unique
  </span>
  <span
    className="array-constraints"
  >
    4-10 items
  </span>
</span>
`;
